### PR TITLE
SOLR-16995 Add a ReplicaCount to keep track of replicas per type

### PR DESCRIPTION
**Pull request by pvcnt**
https://issues.apache.org/jira/browse/SOLR-16995

# Description

Working with replica counts, i.e., the number of replicas per replica type, is currently a bit painful, as the number of replicas is tracked individually for every replica type via individual variables. Solving this issue is also a step towards making it easier to add new replica types.

# Solution

This PR introduces a `ReplicaCount` class (there was one up to Solr 8.x in the autoscaling framework, but it was removed in Solr 9) that is fancy wrapper around a Map<Replica.Type, Integer>.

# Tests

New tests for the `ReplicaCount` class. Existing tests ensure non-regression (there is no added feature).

# Checklist

Please review the following and check all that apply:

- [x] I have reviewed the guidelines for [How to Contribute](https://wiki.apache.org/solr/HowToContribute) and my code conforms to the standards described there to the best of my ability.
- [x] I have created a Jira issue and added the issue ID to my pull request title.
- [x] I have given Solr maintainers [access](https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to contribute to my PR branch. (optional but recommended)
- [x] I have developed this patch against the `main` branch.
- [x] I have run `./gradlew check`.
- [x] I have added tests for my changes.
- [x] I have added documentation for the [Reference Guide](https://github.com/apache/solr/tree/main/solr/solr-ref-guide)

### CODE REVIEW

**Comment by janhoy** [comment]
I think this can benefit from a JIRA issue?

File: solr/solrj/src/java/org/apache/solr/common/cloud/ReplicaCount.java [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+package org.apache.solr.common.cloud;
+
+import java.util.Arrays;
+import java.util.EnumMap;
+import java.util.Locale;
+import java.util.Map;
+import java.util.stream.Collectors;
+import org.apache.solr.common.SolrException;
+
+/** Tracks a number of replica per replica types. This class is mutable. */
```
Comment (sigram) [inline solr/solrj/src/java/org/apache/solr/common/cloud/ReplicaCount.java:26]: Wording seems odd: perhaps "the number of replicas per replica type" ?

Comment (pvcnt) [reply solr/solrj/src/java/org/apache/solr/common/cloud/ReplicaCount.java]: Agreed, that's fixed now, thanks!

**Comment by pvcnt** [comment]
@janhoy You're right, I created one: https://issues.apache.org/jira/browse/SOLR-16995

**Comment by dsmiley** [comment]
Furthermore, PRs titles must start with a JIRA reference (assuming there is one); that's how they are linked.  I'll change that now.

File: solr/core/src/java/org/apache/solr/cloud/api/collections/Assign.java [hunk 2]
```diff
@@ -325,20 +326,18 @@ public static List<ReplicaPosition> getNodesForNewReplicas(
...
-        shard,
-        nrtReplicas,
-        tlogReplicas,
-        pullReplicas,
-        createNodeSet);
+    if (log.isDebugEnabled()) {
+      log.debug(
+          "getNodesForNewReplicas() shard={}, {} , createNodeSet={}",
+          shard,
+          numReplicas.toDebugString(),
```
Comment (dsmiley) [inline solr/core/src/java/org/apache/solr/cloud/api/collections/Assign.java:338]: Might it be reasonable for ReplicaCount.toString() to be this debug string?

**Review by dsmiley — COMMENTED** [review]
+1 Love this!

File: solr/core/src/java/org/apache/solr/cluster/placement/impl/PlacementRequestImpl.java [hunk 3]
```diff
@@ -26,29 +25,23 @@
...
+  private final ReplicaCount numReplicas;
 
   public PlacementRequestImpl(
       SolrCollection solrCollection,
       Set<String> shardNames,
       Set<Node> targetNodes,
-      int countNrtReplicas,
-      int countTlogReplicas,
-      int countPullReplicas) {
+      ReplicaCount numReplicas) {
```
Comment (dsmiley) [inline solr/core/src/java/org/apache/solr/cluster/placement/impl/PlacementRequestImpl.java:40]: Not sure if it's worth adding an overloaded variant for backwards-compatibility, albeit marked deprecated and a comment to remove in 10?

File: solr/solrj/src/java/org/apache/solr/common/cloud/ReplicaCount.java [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+import java.util.Arrays;
+import java.util.EnumMap;
+import java.util.Locale;
+import java.util.Map;
+import java.util.stream.Collectors;
+import org.apache.solr.common.SolrException;
+
+/** Tracks the number of replicas per replica type. This class is mutable. */
+public final class ReplicaCount {
+  private final Map<Replica.Type, Integer> countByType;
```
Comment (dsmiley) [inline solr/solrj/src/java/org/apache/solr/common/cloud/ReplicaCount.java:28]: Should be EnumMap

Comment (pvcnt) [reply solr/solrj/src/java/org/apache/solr/common/cloud/ReplicaCount.java]: Fixed.

File: solr/core/src/java/org/apache/solr/cluster/placement/impl/PlacementRequestImpl.java (hunk 3 above)
Comment (pvcnt) [reply solr/core/src/java/org/apache/solr/cluster/placement/impl/PlacementRequestImpl.java]: Sure, fixed.

File: solr/core/src/java/org/apache/solr/cloud/api/collections/Assign.java (hunk 2 above)
Comment (pvcnt) [reply solr/core/src/java/org/apache/solr/cloud/api/collections/Assign.java]: It sounds reasonable, done.

**Comment by pvcnt** [comment]
I split the changes related to `leaderEligible` into a separate PR: #1911

**Comment by dsmiley** [comment]
HdfsCloudIncrementalBackupTest has been flapping since 2022-08, and lately steadily increasing.

**Pull request merged**